conda/content_trust/__init__.py

```diff
@@ -0,0 +1,238 @@
+    from conda_content_trust.common import (
+        load_metadata_from_file,
+        write_metadata_to_file,
+        SignatureError,
+    )
+    from conda_content_trust.signing import wrap_as_signable
+except ImportError:
+    # SignatureStatus.enabled handles this
+    pass
+
+
+log = getLogger(__name__)
+
+
+class _SignatureVerification:
+    # FUTURE: Python 3.8+, replace with functools.cached_property
+    @property
+    @lru_cache(maxsize=None)
+    def enabled(self):
+        # safety checks must be enabled
+        if not context.extra_safety_checks:
+            return False
+
+        # signing url must be defined
+        if not context.signing_metadata_url_base:
+            log.info(
+                "metadata signature verification requested, "
+                "but no metadata URL base has not been specified."
+            )
+            return False
+
+        # conda_content_trust must be installed
+        try:
+            import conda_content_trust  # noqa: F401
+        except ImportError:
+            log.warn(
+                "metadata signature verification requested, "
+                "but `conda-content-trust` is not installed."
+            )
+            return False
+
+        # create artifact verification directory if missing
+        if not isdir(context.av_data_dir):
+            log.info("creating directory for artifact verification metadata")
+            makedirs(context.av_data_dir)
+
+        # ensure the trusted_root exists
+        if self.trusted_root is None:
+            log.warn("could not find trusted_root data for metadata signature verification")
+            return False
+
+        # ensure the key_mgr exists
+        if self.key_mgr is None:
+            log.warn("could not find key_mgr data for metadata signature verification")
+            return False
+
+        # signature verification is enabled
+        return True
+
+    # FUTURE: Python 3.8+, replace with functools.cached_property
+    @property
+    @lru_cache(maxsize=None)
+    def trusted_root(self):
+        # TODO: formalize paths for `*.root.json` and `key_mgr.json` on server-side
+        trusted = INITIAL_TRUST_ROOT
+
+        # Load current trust root metadata from filesystem
+        for path in sorted(glob(join(context.av_data_dir, "[0-9]*.root.json")), reverse=True):
+            try:
+                int(basename(path).split(".")[0])
+            except ValueError:
+                # prefix is not an int and is consequently an invalid file, skip to the next
+                pass
+            else:
+                log.info(f"Loading root metadata from {path}.")
+                trusted = load_metadata_from_file(path)
+                break
+        else:
+            log.debug(
+                f"No root metadata in {context.av_data_dir}. " "Using built-in root metadata."
+            )
+
+        # Refresh trust root metadata
+        while True:
```

Comment:
Why does this need an endless loop?

Reply:
Because we don't know how many signatures are in the chain of trust? We iteratively keep fetching a newer signature until there are no more to fetch. This could be switched to using a boolean conditional instead of `while True`/`break`.

Reply:
Yeah, let's do the boolean, I'm usually a little worried to see `while True` in parts of code that do network I/O.

Reply:
We discussed this offline and agree that ideally, the signatures would include an additional key, `has_next`, to indicate that there are more signatures to fetch in the chain of trust.